src/parse.rs

```diff
@@ -2,54 +2,78 @@
 
 //! Hex encoding and decoding.
 
+#[cfg(all(feature = "alloc", not(feature = "std")))]
+use alloc::vec::Vec;
 use core::{fmt, str};
 
 use arrayvec::ArrayVec;
 
-#[cfg(all(feature = "alloc", not(feature = "std")))]
-use crate::alloc::vec::Vec;
-use crate::error::InvalidLengthError;
+use crate::error::{InvalidLengthError, OddLengthError};
 use crate::iter::HexToBytesIter;
 
 #[rustfmt::skip]                // Keep public re-exports separate.
-pub use crate::error::{HexToBytesError, HexToArrayError};
+#[doc(inline)]
+pub use crate::error::{HexToVecError, HexToArrayError, InvalidCharError};
 
 /// Trait for objects that can be deserialized from hex strings.
 pub trait FromHex: Sized {
-    /// Error type returned while parsing hex string.
-    type Error: From<HexToBytesError> + Sized + fmt::Debug + fmt::Display;
+    /// Error type returned while constructing type from byte iterator.
+    type FromByteIterError: fmt::Debug + fmt::Display;
+
+    /// Error type returned by `from_hex`.
+    type FromHexError: From<Self::FromByteIterError> + fmt::Debug + fmt::Display;
```

Comment:
Having two errors is quite complicated and I'm not convinced that it achieves anything. In most cases people are bubbling an error up the call stack so one error is better and in those other cases just matching on enum is fine.

I think we should either have:

```rust
enum ParseError<E = Infallible> {
    InvalidChar(InvalidCharError),
    InvalidLength(InvalidLengthError),
    InvalidValue(E),
}
```

Or have `type Error: From<InvalidCharError>`.

I think what we really need though is to decouple invalid char from length error so that invalid length can vary. And I was thinking of having an API that doesn't return `InvalidLength` for guaranteed-valid lengths.